browser/ui/views/infobars/custom_styled_label.cc

```diff
@@ -12,6 +12,24 @@
 
 CustomStyledLabel::~CustomStyledLabel() = default;
 
+void CustomStyledLabel::Layout() {
+  // Link click doesn't work when StyledLabel is multilined.
+  // And there is upstream bug -
+  // https://bugs.chromium.org/p/chromium/issues/detail?id=1371538 Below code is
+  // copied from WIP
+  // PR(https://chromium-review.googlesource.com/c/chromium/src/+/3934027) to
+  // skip layout when it doesn't need layout. Layout() happens between
+  // OnMousePressed() and OnMouseReleased(). As Layout() deletes all children,
```

Comment:
Whoa, this must have been hard to find out! Well spotted!

Layout() seems to be called whenever mouse hovers and exits the LinkFragment. And it causes SetFontList() even it's already underlined, which seems to be a bug. Resetting font list triggers full cycle of layout from the top of browser view tree. So following code seems to make it not to relayout everyting.

```diff
void LinkFragment::RecalculateFont() {
  // Check whether any link fragment should be underlined.
  bool should_be_underlined = IsUnderlined();
  for (LinkFragment* current_fragment = next_fragment_;
       !should_be_underlined && current_fragment != this;
       current_fragment = current_fragment->next_fragment_) {
    should_be_underlined = current_fragment->IsUnderlied();
  }

  // If the style differs from the current one, update.
  if ((font_list().GetFontStyle() & gfx::Font::UNDERLINE) !=
      should_be_underlined) {
    auto MaybeUpdateStyle = [should_be_underlined](LinkFragment* fragment) {
      const int style = fragment->font_list().GetFontStyle();
      const int intended_style = should_be_underlined
                                     ? (style | gfx::Font::UNDERLINE)
                                     : (style & ~gfx::Font::UNDERLINE);
+      if (style == intended_style)
+        return;
      fragment->Label::SetFontList(
          fragment->font_list().DeriveWithStyle(intended_style));
      fragment->SchedulePaint();
    };
```

And a potential problem with the current code could be when we set a new font to a fragment, then it won't be laid out without explictly changing the size of the fragment. But now that this is not our case, I'm okay with the current fix. It's up to you! But could you update comments here?

Reply:
Thanks for digging more. I added more comments about it and let's wait fixing from upstream.
One thing I'm still curious thing is this doesn't happen when StyleLable is not multilined.
Needs to investigate more later :)

Reply:
> doesn't happen when StyleLable is not multilined.

Yeah, there seems to be multiple LinkFragments, and it doesn't look working well :(

Reply:
Ah, I see. it's `LinkFragment` issue.